File: simulacrum/src/main/java/io/github/projectet/dmlSimulacrum/util/Animation.java
package io.github.projectet.dmlSimulacrum.util;

public class Animation {
    private boolean hasStarted = false;
    private int renderPos = 0;
    private int renderStop = 0;
    private int currentTick = 0;
    private long lastAlterTick = 0;
    private String currentString = "";

    public String animate(String string, int totalTickTime, long currentWorldTick, boolean loop) {
        if(lastAlterTick != currentWorldTick) {
            alterString(string, totalTickTime, loop);
            lastAlterTick = currentWorldTick;
        }

        return currentString;
    }

    public void alterString(String string, int totalTickTime, boolean loop) {
        if(!hasStarted) {
            this.hasStarted = true;
            this.renderPos = 0;
            this.renderStop = string.length();
        } else {
            currentTick++;
            if(renderPos < renderStop) {
                if(renderPos >= 0) {
                    this.currentString = string.substring(0, renderPos);
                }
                this.renderPos = currentTick % totalTickTime == 0 ? renderPos + 1 : renderPos;
            } else {
                if(loop) {
                    this.hasStarted = currentTick % totalTickTime != 0;
                }
                this.currentString = string;
            }
        }
    }

    public boolean hasFinished() {
        return renderPos == renderStop && renderStop != 0;
    }

    public void clear() {
        this.currentString = "";
        this.hasStarted = false;
        this.renderPos = 0;
        this.renderStop = 0;
        this.currentTick = 0;
    }
}

File: simulacrum/src/main/kotlin/dev/nathanpb/dml/simulacrum/util/Animation.kt
package dev.nathanpb.dml.simulacrum.util

class Animation {

    private var hasStarted = false
    private var renderPos = 0
    private var renderStop = 0
    private var currentTick = 0
    private var lastAlterTick: Long = 0
    private var currentString = ""


    fun animate(key: String, totalTickTime: Int, currentWorldTick: Long, loop: Boolean): String {
        return animate(key, totalTickTime, currentWorldTick, loop, -1)
    }

    fun animate(key: String, totalTickTime: Int, currentWorldTick: Long, loop: Boolean, intArg: Int): String {
        if (lastAlterTick != currentWorldTick) {
            alterString(key, totalTickTime, loop)
            lastAlterTick = currentWorldTick
        }
        return currentString
    }

    fun alterString(string: String, totalTickTime: Int, loop: Boolean) {
        if (!hasStarted) {
            hasStarted = true
            renderPos = 0
            renderStop = string.length
        } else {
            currentTick++
            if (renderPos < renderStop) {
                if (renderPos >= 0) {
                    currentString = string.substring(0, renderPos)
                }
                renderPos = if (currentTick % totalTickTime == 0) renderPos + 1 else renderPos
            } else {
                if (loop) {
                    hasStarted = currentTick % totalTickTime != 0
                }
                currentString = string
            }
        }
    }

    fun hasFinished(): Boolean {
        return renderPos == renderStop && renderStop != 0
    }

    fun clear() {
        currentString = ""
        hasStarted = false
        renderPos = 0
        renderStop = 0
        currentTick = 0
    }
}